File: sdks/php/src/ApiException.php
<?php

declare(strict_types=1);

namespace HookSniff;

/**
 * HookSniff API Exception
 */
class ApiException extends \RuntimeException
{
    /** @var array<string, string> */
    public array $headers;

    /**
     * @param mixed $body
     * @param array<string, string> $headers
     */
    public function __construct(
        public readonly int $code,
        public readonly mixed $body,
        array $headers = [],
    ) {
        $message = sprintf('HookSniff API Error %d: %s', $code, is_string($body) ? $body : json_encode($body));
        parent::__construct($message, $code);
        $this->headers = $headers;
    }
}

File: sdks/csharp/src/HookSniff/ApiException.cs
namespace HookSniff;

/// <summary>
/// HookSniff API Exception
/// </summary>
public class ApiException : Exception
{
    public int StatusCode { get; }
    public object? Body { get; }
    public Dictionary<string, string> Headers { get; }

    public ApiException(int statusCode, object? body, Dictionary<string, string>? headers = null)
        : base($"HookSniff API Error {statusCode}: {body}")
    {
        StatusCode = statusCode;
        Body = body;
        Headers = headers ?? new Dictionary<string, string>();
    }
}
